### PR TITLE
Initial try at deduping event listeners.

### PR DESCRIPTION
**Pull request by tkh44**
Uses the registration technique from react-virtualized.

I'm not very experienced with flow so I may have done some things wrong.

#1

### CODE REVIEW

File: src/eventRegistration.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
+// @flow
+type EventHandlers = {
+  (onMouseMove): (MouseEvent) => void,
+  (onWindowResize): (Event) => void
+};
+
+type MountedInstance = {
+  component: React$Element<*>,
+  handlers: Object
```
Comment (tkh44) [inline src/eventRegistration.js:9]: I don't know how to get this type working. Any feedback is appreciated.

**Comment by aweary** [comment]
@tkh44 thanks for the PR, this looks great! What if we instead provided an API for this instead of providing a default solution? Instead of always deduping listeners (which might actually be more expensive for 1-2 instances) we add an optional prop  so users can pass in their own implementation (so deduping behavior can be app-specific).

Then we could provide a default implementation as a separate export, so only users who want to use it pay the bundle size cost

```jsx
import React from 'react'
import Perimeter from 'react-perimeter'
import eventManager from 'react-perimeter/eventManager'
//...
<Perimeter eventManager={eventManager} />
```

`eventManager` would have a documented API (`registerListeners`, `removeListeners`) so anyone could implement their own. If `eventManager` isn't provided we just add the listener directly to `window` like we do now.

What do you think?

**Comment by tkh44** [comment]
It sounds like a good idea, but I'm just wondering what the benefit would be compared to the added complexity (from user perspective). It seems like it would add a lot of moving parts to a beautifully simple api.

**Comment by jnsdls** [comment]
I agree that it's quite a bit of moving parts if you want to make it conditional and let people pass their own register & remove listeners functions. (I like the customizability aspect but I'm not sure if it's necessary, we should just come up with the 90 percentile best solution and go with that.)

In my proposed solution from #1 we could handle the "global vs non-global" problem very simply by the way:

1) check if there is a MouseMoveProvider on the context
2) if yes -> subscribe || if no -> use own window.eventListener

and then if we wanted to go super deep we could still do the eventManager as a potential prop to <MouseMoveProvider /> (think history in React-Router)

**Comment by aweary** [comment]
Closing in favor of https://github.com/aweary/react-perimeter/pull/14 🎉

**Pull request closed without merging**